File: project_euler/problem_190/solution.py
"""
Problem 190: Maximising a Weighted Product

For m from 2 to 15, find the maximum of prod(x_i^i) subject to sum(x_i) = m,
then sum floor of each maximum.
"""
import math
from decimal import Decimal, getcontext

def solve():
    getcontext().prec = 50
    total = 0

    for m in range(2, 16):
        # Optimal: x_i = 2i / (m+1)
        # P_m = product of (2i/(m+1))^i
        log_pm = Decimal(0)
        for i in range(1, m + 1):
            xi = Decimal(2 * i) / Decimal(m + 1)
            log_pm += i * xi.ln()

        pm = log_pm.exp()
        total += int(pm)

    print(total)

if __name__ == "__main__":
    solve()

File: project_euler/problem_190/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    long long total = 0;

    for (int m = 2; m <= 15; m++) {
        // x_i = 2i / (m+1)
        // P_m = product of (2i/(m+1))^i for i=1..m
        double log_pm = 0.0;
        for (int i = 1; i <= m; i++) {
            log_pm += i * log(2.0 * i / (m + 1));
        }
        long long pm = (long long)floor(exp(log_pm));
        total += pm;
    }

    cout << total << endl;
    return 0;
}
